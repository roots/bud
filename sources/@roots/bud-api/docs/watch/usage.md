---
title: Usage
---

To reload the page automatically when a particular file or directory is modified, you can use bud.watch.

All paths are relative to the project root directory:

```typescript title=bud.config.ts
bud.watch([
  `./app/config`,
  `./src/views`,
])
```

It is generally recommend to set file paths using [bud.path](/reference/bud.path):

```typescript title=bud.config.ts
bud.watch([
  bud.path(`@src/pages/page.php`)
])
```

You can use [bud.glob](/reference/bud.glob) to specify paths using pattern matching syntax:

```typescript title=bud.config.ts
bud.watch([
  bud.path(`@src/pages/**/*`)
])
```

### Setting options

All options from [chokidar](https://github.com/paulmillr/chokidar) are supported. In the future we will likely switch to node's built in file watch API, but most options are cross compatible.

Setting watch options can be accomplished in a couple ways.

You can set options along with the file paths you want to watch by passing a second parameter:

```typescript title=bud.config.ts
bud.watch([bud.path(`@src/pages`)], {
  interval: 1000,
  usePolling: true,
})
```

If you only want to set options, you can also choose to just pass the options directly:

```typescript title=bud.config.ts
bud.watch({
  interval: 1000,
  usePolling: true,
})
```
