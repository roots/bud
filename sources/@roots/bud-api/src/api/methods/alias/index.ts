export type {facade} from './alias.interface'
export {alias as method} from './alias.method'
export {alias} from './alias.method'
