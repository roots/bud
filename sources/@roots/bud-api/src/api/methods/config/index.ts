import type {Framework} from '@roots/bud-framework'
import Webpack from 'webpack'

export interface config {
  (overrides: Partial<Webpack.Configuration>): Framework
}

export function config(
  overrides: (
    config: Partial<Webpack.Configuration>,
  ) => Partial<Webpack.Configuration>,
): Framework {
  const ctx = this as Framework

  if (!overrides)
    throw new Error(
      'config overrides must pass a callback function that returns a webpack configuration',
    )

  ctx.hooks.action('event.build.after', async app => {
    app.build.config = overrides(app.build.config)
  })

  return ctx
}
