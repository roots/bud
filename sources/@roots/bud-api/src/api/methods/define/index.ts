import type {Bud} from '@roots/bud-framework'
import type {DefinePlugin} from 'webpack'

export interface define {
  (this: Bud, values: DefinePlugin['definitions']): Bud
}

export function define(
  this: Bud,
  values: DefinePlugin['definitions'],
): Bud {
  this.extensions.get('webpack-define-plugin').mutateOptions(options => {
    Object.entries(values).forEach(([k, v]) => {
      options.set(k, v)
    })

    return options
  })

  return this
}
