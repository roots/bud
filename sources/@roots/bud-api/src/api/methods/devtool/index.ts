import type {Framework} from '@roots/bud-framework'
import type {Configuration} from 'webpack'

export interface devtool {
  (devtool?: Configuration['devtool']): Promise<Framework>
}

export interface facade {
  (devtool?: Configuration['devtool']): Framework
}

export const devtool: devtool = async function (devtool = false) {
  this as Framework

  const value = devtool ?? 'cheap-module-source-map'

  this.hooks.on('build.devtool', value)

  this.api.log('success', {prefix: 'devtool', message: devtool})

  return this
}
