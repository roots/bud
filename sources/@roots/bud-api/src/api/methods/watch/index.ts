import type {Framework} from '@roots/bud-framework'
import {chokidar} from '@roots/bud-support'

export interface watch {
  (
    /**
     * Watched files
     */
    files: Array<string>,
    /**
     * Watcher options
     */
    options?: chokidar.WatchOptions,
  ): Framework
}

export const watch: watch = function (files, options = {}) {
  const ctx = this as Framework

  files = Array.isArray(files) ? files : [files]

  ctx.hooks
    .on('dev.watch.files', fileSet => {
      files.forEach(file => fileSet.add(file))
      return fileSet
    })
    .hooks.on('dev.watch.options', hookValue => ({
      ...(hookValue ?? {}),
      ...(options ?? {}),
    }))

  return ctx
}
