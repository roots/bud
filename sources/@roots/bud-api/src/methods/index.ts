export {alias} from './alias/index.js'
export {assets, assets as copy} from './assets/index.js'
export {bundle} from './bundle/index.js'
export {
  config,
  config as webpackConfig,
  config as override,
} from './config/index.js'
export {copyDir} from './copyDir/index.js'
export {copyFile} from './copyFile/index.js'
export {define} from './define/index.js'
export {devtool} from './devtool/index.js'
export {entry} from './entry/index.js'
export {experiments} from './experiments/index.js'
export {externals} from './externals/index.js'
export {hash, hash as version} from './hash/index.js'
export {html, html as template} from './html/index.js'
export {minimize} from './minimize/index.js'
export {persist} from './persist/index.js'
export {provide} from './provide/index.js'
export {proxy} from './proxy/index.js'
export {compilePaths} from './compilePaths/index.js'
export {runtime} from './runtime/index.js'
export {serve} from './serve/index.js'
export {setPublicUrl} from './setPublicUrl/index.js'
export {setProxyUrl} from './setProxyUrl/index.js'
export {setPublicProxyUrl} from './setPublicProxyUrl/index.js'
export {setUrl} from './setUrl/index.js'
export {splitChunks} from './splitChunks/index.js'
export {use} from './use/index.js'
export {watch} from './watch/index.js'
