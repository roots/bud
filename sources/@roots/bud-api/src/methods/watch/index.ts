import type {Bud} from '@roots/bud-framework'
import type {WatchOptions} from '@roots/bud-support/chokidar'

import isString from '@roots/bud-support/lodash/isString'

export type Parameters =
  | [Array<string> | string, WatchOptions?]
  | [WatchOptions]

export interface watch {
  (...params: Parameters): Promise<Bud>
}

export const watch: watch = async function (this: Bud, ...params) {
  if (!this.isDevelopment) return this

  const [files, options] = makeFilesAndOptions(...params)

  this.hooks.on(`dev.watch.files`, (watchlist = new Set()) => {
    files.map(file => watchlist.add(file))
    return watchlist
  })

  this.hooks.on(`dev.watch.options`, (watchOptions = {}) => ({
    ...watchOptions,
    ...options,
  }))

  return this
}

const makeFilesAndOptions = (
  ...params: Parameters
): [Array<string>, WatchOptions] => {
  let files: Array<string> = []
  let options: WatchOptions = {}

  if (Array.isArray(params[0])) {
    files = params[0]
  }
  if (isString(params[0])) {
    files = [params[0]]
  }
  if (params[0] && !Array.isArray(params[0]) && !isString(params[0])) {
    options = params[0]
  }
  if (params[1]) {
    options = params[1]
  }

  return [files, options]
}
