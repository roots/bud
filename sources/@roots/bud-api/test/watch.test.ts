import {Bud, factory} from '@repo/test-kit'
import {watch as subject} from '@roots/bud-api/methods/watch'
import {beforeEach, describe, expect, it, vi} from 'vitest'

describe(`bud.watch`, () => {
  describe(`in development`, () => {
    let bud: Bud
    let watch: subject

    beforeEach(async () => {
      bud = await factory({mode: `development`})
      watch = subject.bind(bud)
    })

    it(`should add watch files`, async () => {
      await watch(`1/*.js`)

      const value = bud.hooks.filter(`dev.watch.files`)

      if (!(value instanceof Set))
        throw new Error(`watch files should be a set`)

      expect(Array.from(value)).toMatchSnapshot(
        expect.arrayContaining([`1/*.js`]),
      )
    })
  })

  describe(`in production`, () => {
    let bud: Bud
    let watch: subject

    beforeEach(async () => {
      bud = await factory({mode: `production`})
      watch = subject.bind(bud)
    })

    it(`should not run`, async () => {
      expect(bud.mode).toBe(`production`)

      const watch = subject.bind(bud)

      await watch(`1/*.js`)

      const value = bud.hooks.filter(`dev.watch.files`)
      expect(value).toBeUndefined()
    })
  })

  describe(`string file`, () => {
    let bud: Bud
    let watch: subject

    beforeEach(async () => {
      bud = await factory({mode: `development`})
      watch = subject.bind(bud)
    })

    it(`should add files`, async () => {
      await watch(`1/*.js`)

      // @ts-ignore
      expect(Array.from(bud.hooks.filter(`dev.watch.files`))).toEqual(
        expect.arrayContaining([`1/*.js`]),
      )
    })
  })

  describe(`array files`, () => {
    let bud: Bud
    let watch: subject

    beforeEach(async () => {
      bud = await factory({mode: `development`})
      watch = subject.bind(bud)
    })

    it(`should add files`, async () => {
      await watch([`1/*.js`, `2/*.js`])

      // @ts-ignore
      expect(Array.from(bud.hooks.filter(`dev.watch.files`))).toEqual(
        expect.arrayContaining([`1/*.js`, `2/*.js`]),
      )
    })
  })

  describe(`[Array, Options] parameters`, () => {
    let bud: Bud
    let watch: subject

    beforeEach(async () => {
      bud = await factory({mode: `development`})
      watch = subject.bind(bud)
    })

    it(`should add watch files and options`, async () => {
      await watch([`1/*.js`], {
        usePolling: true,
      })

      const value = bud.hooks.filter(`dev.watch.files`)

      if (!(value instanceof Set))
        throw new Error(`watch files should be a set`)

      expect(Array.from(value)).toMatchSnapshot(
        expect.arrayContaining([`1/*.js`]),
      )

      const options = bud.hooks.filter(`dev.watch.options`)
      expect(options).toMatchInlineSnapshot(`
        {
          "usePolling": true,
        }
      `)
    })
  })

  describe(`Options only`, () => {
    let bud: Bud
    let watch: subject

    beforeEach(async () => {
      bud = await factory({mode: `development`})
      watch = subject.bind(bud)
    })

    it(`should add options`, async () => {
      await watch({
        usePolling: true,
      })

      const options = bud.hooks.filter(`dev.watch.options`)
      expect(options).toMatchInlineSnapshot(`
        {
          "usePolling": true,
        }
      `)
    })
  })
})
