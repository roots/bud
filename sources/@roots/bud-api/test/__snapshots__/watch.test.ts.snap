// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bud.watch > [Array, Options] parameters > should add watch files and options 1`] = `
[
  "1/*.js",
]
`;

exports[`bud.watch > in development > adds watch files 1`] = `
[
  "1/*.js",
]
`;

exports[`bud.watch > in development > should add watch files 1`] = `
[
  "1/*.js",
]
`;
